### PR TITLE
Ofmcc 375 send documents to crm

### PR DESCRIPTION
**Pull request by vietle-cgi**

### CODE REVIEW

**Review by weskubo-cgi — CHANGES_REQUESTED** [review]

File: frontend/src/stores/documents.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  documents?.forEach((document, index) => {
+    fileMapping.push({
+      ofm_subject: document.file[0]?.name,
+      ofm_extension: getFileExtension(document.file[0]?.name),
+      ofm_file_size: document.file[0]?.size,
+      ofm_description: document.description,
+      entity_name_set: document.entityName,
+      regardingid: document.entityId,
+    })
+    payload.append(`document[${index}]`, document.file[0])
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply frontend/src/stores/documents.js]: this is updated

File: frontend/src/stores/documents.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import { defineStore } from 'pinia'
+import { isEmpty } from 'lodash'
+
+function getFileExtension(fileName) {
+  if (fileName) return fileName.slice(fileName.lastIndexOf('.')).toLowerCase()
+  return ''
+}
+
+function mapDocumentFileObjectForBack(documents) {
+  let payload = new FormData()
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply frontend/src/stores/documents.js]: this is updated

File: frontend/src/stores/documents.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+import { ApiRoutes } from '@/utils/constants'
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply frontend/src/stores/documents.js]: I removed this document store and use documentService instead

File: frontend/src/components/ui/AppDocumentUpload.vue [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      fileAccept: ['image/png', 'image/jpeg', 'image/jpg', '.pdf', '.png', '.jpg', '.jpeg', '.heic', '.doc', '.docx', '.xls', '.xlsx'],
+      fileExtensionAccept: ['pdf', 'png', 'jpg', 'jpeg', 'heic', 'doc', 'docx', 'xls', 'xlsx'],
+      fileFormats: 'PDF, JPEG, JPG, PNG, HEIC, DOC, DOCX, XLS and XLSX',
+      fileInputError: [],
+      fileMap: new Map(),
+      fileRules: [],
+      editedIndex: -1,
+      defaultItem: {
+        entityName: this.entityName,
+      },
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply frontend/src/components/ui/AppDocumentUpload.vue]: this has been removed

File: frontend/src/components/ui/AppDocumentUpload.vue [hunk 5]
```diff
@@ -0,0 +1,91 @@
+<template>
+  <v-container fluid class="pa-0">
+    <v-form ref="form" v-model="isValidForm">
+      <div>The maximum file size is 4MB for each document. Accepted file types are jpg, jpeg, heic, png, pdf, docx, doc, xls, and xlsx.</div>
+      <div v-if="documents.length > 0" class="mt-6">
+        <v-row v-for="(item, index) in documents" :key="index" no-gutters>
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply frontend/src/components/ui/AppDocumentUpload.vue]: I updated the code to use uuid

**Review by weskubo-cgi — CHANGES_REQUESTED** [review]

File: backend/src/components/utils.js [hunk 6]
```diff
@@ -201,6 +201,21 @@ async function patchOperationWithObjectId(operation, objectId, payload) {
...
+  const url = config.get('dynamicsApi:apiEndpoint') + '/api/documents'
+  log.info('postDocuments Url', url)
+
+  try {
+    const response = await axios.post(url, payload, { headers: headers })
+    logResponse('postDocuments', response)
+    return response.data
+  } catch (e) {
+    log.error('postDocuments Error', e)
+    log.error('postDocuments Error', e.response ? e.response.status : e.message)
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply backend/src/components/utils.js]: I removed 1 of them

File: frontend/src/services/documentService.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      formData.append(`file${index}`, file)
+    }
+  })
+  formData.append('fileMapping', JSON.stringify(fileMapping))
+  return formData
+}
+
+export default {
+  async createDocuments(documents) {
+    if (!isEmpty(documents)) {
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply frontend/src/services/documentService.js]: I updated it to use early return pattern

File: frontend/src/services/documentService.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import { ApiRoutes } from '@/utils/constants'
+import ApiService from '@/common/apiService'
+import { isEmpty } from 'lodash'
+
+function mapDocumentFileObjectForBack(documents) {
+  let formData = new FormData()
+  let fileMapping = []
+  documents?.forEach((document, index) => {
+    if (document.file?.length > 0) {
```
[the remark this answers is not in the document]
Comment (vietle-cgi) [reply frontend/src/services/documentService.js]: I updated it to use early return pattern

**Review by weskubo-cgi — APPROVED** [review]

**Pull request merged**